File: client/src/app/services/friends/friends.service.ts
import { Injectable } from '@angular/core';
import { AuthenticationService } from '@app/services/authentication/authentication.service';
import { SocketHandlerService } from '@app/services/socket-handler/socket-handler.service';
import { FriendsEvents } from '@common/events/friends.events';
import { UserIdName } from '@common/interfaces/user-id-name';

@Injectable({
    providedIn: 'root',
})
export class FriendsService {
    allUsers: UserIdName[] = [];
    friends: UserIdName[] = [];
    pendingRequests: UserIdName[] = [];
    sentRequests: UserIdName[] = [];
    searchResults: UserIdName[] = [];
    currentQuery: string = '';

    allDataLoaded: boolean = false;

    constructor(
        private readonly authService: AuthenticationService,
        private readonly socketHandler: SocketHandlerService,
    ) {}

    returnAllData() {
        this.allDataLoaded = false;
        this.onReturnUsers();
        this.socketHandler.send(FriendsEvents.ReturnAllData, this.authService.userId);
    }

    onReturnUsers() {
        let loadedDataTypes = 0;
        const totalDataTypes = 4;

        const checkAllDataLoaded = () => {
            loadedDataTypes++;
            if (loadedDataTypes >= totalDataTypes) {
                this.allDataLoaded = true;
            }
        };

        this.socketHandler.on(FriendsEvents.ReturnAllUsers, (data: UserIdName[]) => {
            this.allUsers = data;
            // this.searchResults = this.allUsers.filter((user) => !this.isFriend(user));
            this.searchUsers(this.currentQuery);
            checkAllDataLoaded();
        });
        this.socketHandler.on(FriendsEvents.ReturnAllFriends, (data: UserIdName[]) => {
            this.friends = data;
            checkAllDataLoaded();
        });
        this.socketHandler.on(FriendsEvents.ReturnAllPendingRequests, (data: UserIdName[]) => {
            this.pendingRequests = data;
            checkAllDataLoaded();
        });
        this.socketHandler.on(FriendsEvents.ReturnAllSentRequests, (data: UserIdName[]) => {
            this.sentRequests = data;
            checkAllDataLoaded();
        });
    }

    stopReturningUsers() {
        this.socketHandler.socket.removeListener(FriendsEvents.ReturnAllUsers);
        this.socketHandler.socket.removeListener(FriendsEvents.ReturnAllFriends);
        this.socketHandler.socket.removeListener(FriendsEvents.ReturnAllPendingRequests);
        this.socketHandler.socket.removeListener(FriendsEvents.ReturnAllSentRequests);
        this.currentQuery = '';
    }

    sendFriendRequest(toUserId: string) {
        this.socketHandler.send(FriendsEvents.RequestSent, { user: this.authService.userId, friend: toUserId });
    }

    acceptFriendRequest(friendId: string) {
        this.socketHandler.send(FriendsEvents.RequestAccepted, { user: this.authService.userId, friend: friendId });
    }

    rejectFriendRequest(friendId: string) {
        this.socketHandler.send(FriendsEvents.RequestRejected, { user: this.authService.userId, friend: friendId });
    }

    cancelRequest(friendId: string) {
        this.socketHandler.send(FriendsEvents.RequestCanceled, { user: this.authService.userId, friend: friendId });
    }

    removeFriend(friendId: string) {
        this.socketHandler.send(FriendsEvents.FriendRemoved, { user: this.authService.userId, friend: friendId });
    }

    searchUsers(query: string) {
        this.currentQuery = query;
        if (!query.trim()) {
            this.searchResults = this.allUsers.filter((user) => !this.isFriend(user));
        } else {
            const q = query.toLowerCase();
            this.searchResults = this.allUsers.filter((user) => (user.name || '').toLowerCase().includes(q) && !this.isFriend(user));
        }
    }

    isFriend(user: UserIdName): boolean {
        return this.friends.some((f) => f.id === user.id);
    }

    isRequestPending(user: UserIdName): boolean {
        return this.pendingRequests.some((r) => r.id === user.id);
    }

    isRequestSent(user: UserIdName): boolean {
        return this.sentRequests.some((r) => r.id === user.id);
    }

    isEligible(user: UserIdName): boolean {
        return !this.isFriend(user) && !this.isRequestPending(user) && !this.isRequestSent(user);
    }
}

File: client-mobile/app/src/main/java/com/example/polyquiz/friends/domain/FriendsService.kt
package com.example.polyquiz.friends.domain

import com.example.polyquiz.auth.domain.UserIdName
import com.example.polyquiz.constants.ChoiceInfo
import com.example.vanillaprototype.socket.SocketHandler
import com.example.polyquiz.constants.FriendsEvents
import com.google.firebase.perf.FirebasePerformance
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import org.json.JSONObject

class FriendsService {
    private val _allUsers = MutableStateFlow(listOf<UserIdName>())
    val allUsers: StateFlow<List<UserIdName>> get() = _allUsers

    private val _friends = MutableStateFlow(listOf<UserIdName>())
    val friends: StateFlow<List<UserIdName>> get() = _friends

    private val _pendingRequests = MutableStateFlow(listOf<UserIdName>())
    val pendingRequests: StateFlow<List<UserIdName>> get() = _pendingRequests

    private val _sentRequests = MutableStateFlow(listOf<UserIdName>())
    val sentRequests: StateFlow<List<UserIdName>> get() = _sentRequests

    private val _searchResults = MutableStateFlow<List<UserIdName>>(emptyList())
    val searchResults: StateFlow<List<UserIdName>> get() = _searchResults

    private val _allDataLoaded = MutableStateFlow(false)
    val allDataLoaded: StateFlow<Boolean> = _allDataLoaded

    private val mSocket = SocketHandler.getSocket()
    private var userId: String = ""

    fun initialize(userId: String) {

        this.userId = userId
    }

    fun returnAllData() {
        _allDataLoaded.value = false
        onReturnUsers();
        mSocket.emit(FriendsEvents.RETURN_ALL_DATA.value, userId);
    }

    fun onReturnUsers() {

        var loadedDataTypes = 0
        val totalDataTypes = 4

        fun checkLoadedDataTypes() {
            loadedDataTypes++
            if (loadedDataTypes == totalDataTypes) {
                _allDataLoaded.value = true
            }
        }

        mSocket.on(FriendsEvents.RETURN_ALL_USERS.value) { args: Array<Any> ->
            if (args.isNotEmpty()) {
                val type = object : TypeToken<List<UserIdName>>() {}.type
                val data: List<UserIdName> = Gson().fromJson(args[0].toString(), type)
                _allUsers.value = data
                _searchResults.value = _allUsers.value
                checkLoadedDataTypes()
            }
        }
        mSocket.on(FriendsEvents.RETURN_ALL_FRIENDS.value) { args: Array<Any> ->
            if (args.isNotEmpty()) {
                val type = object : TypeToken<List<UserIdName>>() {}.type
                val data: List<UserIdName> = Gson().fromJson(args[0].toString(), type)
                _friends.value = data
                checkLoadedDataTypes()
            }
        }
        mSocket.on(FriendsEvents.RETURN_ALL_PENDING_REQUESTS.value) { args: Array<Any> ->
            if (args.isNotEmpty()) {
                val type = object : TypeToken<List<UserIdName>>() {}.type
                val data: List<UserIdName> = Gson().fromJson(args[0].toString(), type)
                _pendingRequests.value = data
                checkLoadedDataTypes()
            }
        }
        mSocket.on(FriendsEvents.RETURN_ALL_SENT_REQUESTS.value) { args: Array<Any> ->
            if (args.isNotEmpty()) {
                val type = object : TypeToken<List<UserIdName>>() {}.type
                val data: List<UserIdName> = Gson().fromJson(args[0].toString(), type)
                _sentRequests.value = data
                checkLoadedDataTypes()
            }
        }
    }

    fun stopReturningUsers() {
        mSocket.off(FriendsEvents.RETURN_ALL_USERS.value)
        mSocket.off(FriendsEvents.RETURN_ALL_FRIENDS.value)
        mSocket.off(FriendsEvents.RETURN_ALL_PENDING_REQUESTS.value)
        mSocket.off(FriendsEvents.RETURN_ALL_SENT_REQUESTS.value)
    }

    fun sendFriendRequest(toUserId: String) {
        val friendInfos = JSONObject(Gson().toJson(FriendsInfo(userId, toUserId)))
        mSocket.emit(FriendsEvents.REQUEST_SENT.value, friendInfos)
    }

    fun acceptFriendRequest(friendId: String) {
        val friendInfos = JSONObject(Gson().toJson(FriendsInfo(userId, friendId)))
        mSocket.emit(FriendsEvents.REQUEST_ACCEPTED.value, friendInfos)
    }

    fun rejectFriendRequest(friendId: String) {
        val friendInfos = JSONObject(Gson().toJson(FriendsInfo(userId, friendId)))
        mSocket.emit(FriendsEvents.REQUEST_REJECTED.value, friendInfos)
    }

    fun cancelRequest(friendId: String) {
        val friendInfos = JSONObject(Gson().toJson(FriendsInfo(userId, friendId)))
        mSocket.emit(FriendsEvents.REQUEST_CANCELED.value, friendInfos)
    }

    fun removeFriend(friendId: String) {
        val friendInfos = JSONObject(Gson().toJson(FriendsInfo(userId, friendId)))
        mSocket.emit(FriendsEvents.FRIEND_REMOVED.value, friendInfos)
    }

}
